File: App/Resources/AlgorithmDetails/dualpivotquicksort.bundle/dualpivotquicksort.cs
using System;

public class DualPivotQuickSort {
  public static (int, int) Partition(int[] array, int low, int high) {
    if (array[low] > array[high]) {
      (array[low], array[high]) = (array[high], array[low]);
    }
    var j = low + 1;
    var g = high - 1;
    var k = low + 1;
    var p = array[low];
    var q = array[high];
    while (k <= g) {
      if (array[k] < p) {
        (array[k], array[j]) = (array[j], array[k]);
        j++;
      } else if (array[k] >= q) {
        while (array[g] > q && k < g) {
          g--;
        }
        (array[k], array[g]) = (array[g], array[k]);
        g--;
        if (array[k] < p) {
          (array[k], array[j]) = (array[j], array[k]);
          j++;
        }
      }
      k++;
    }
    j--;
    g++;
    (array[low], array[j]) = (array[j], array[low]);
    (array[high], array[g]) = (array[g], array[high]);
    return (j, g);
  }

  public static int[] Sort(int[] array, int low, int high) {
    if (low < high) {
      var (j, g) = Partition(array, low, high);
      Sort(array, low, j - 1);
      Sort(array, j + 1, g - 1);
      Sort(array, g + 1, high);
    }
    return array;
  }

  public static void Main(String[] args) {
    int[] array = {0, 39, 21, 62, 91, 77, 14, 23,
      90, 69, 51, 81, 68, 83, 32, 56};
    Sort(array, 0, array.Length - 1);
    string result = "[" + String.Join(", ", array) + "]";
    Console.WriteLine(result);
  }
}

File: App/Resources/AlgorithmDetails/dualpivotquicksort/dualpivotquicksort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int *a, int *b) {
  int t = *a;
  *a = *b;
  *b = t;
}

void printList(int arr[], int n) {
  for (int i = 0; i < n; i++) {
    if (i == 0) {
      printf("[%d, ", arr[i]);
    } else if (i != n - 1) {
      printf("%d, ", arr[i]);
    } else {
      printf("%d]", arr[i]);
    }
  }
}

void partition(int arr[], int low, int high, int *outJ, int *outG) {
  if (arr[low] > arr[high]) {
    swap(&arr[low], &arr[high]);
  }
  int j = low + 1;
  int g = high - 1;
  int k = low + 1;
  int p = arr[low];
  int q = arr[high];
  while (k <= g) {
    if (arr[k] < p) {
      swap(&arr[k], &arr[j]);
      j++;
    } else if (arr[k] >= q) {
      while (arr[g] > q && k < g) {
        g--;
      }
      swap(&arr[k], &arr[g]);
      g--;
      if (arr[k] < p) {
        swap(&arr[k], &arr[j]);
        j++;
      }
    }
    k++;
  }
  j--;
  g++;
  swap(&arr[low], &arr[j]);
  swap(&arr[high], &arr[g]);
  *outJ = j;
  *outG = g;
}

void sort(int arr[], int low, int high) {
  if (low < high) {
    int j, g;
    partition(arr, low, high, &j, &g);
    sort(arr, low, j - 1);
    sort(arr, j + 1, g - 1);
    sort(arr, g + 1, high);
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, 0, size - 1);
  printList(array, size);
  return 0;
}
